Add new rule `no-shadow-route-definition`

raido opened the pull request:
This PR adds new rule to avoid route path definition collisions for Router.

Example message: `Route "blog" (/blog, 3L:8C) is shadowing route "blog" (/blog, 2L:1C)`

Closes: #951

raido commented on tests/lib/rules/no-shadow-route-definition.js:
```diff
@@ -0,0 +1,198 @@
+    this.route("post")`,
+
+    // With dynamic segment:
+    'this.route("blog", { path: ":blog" });',
```
Actually this is incorrect for this rule, as both path definitions would equal `blog/<value>` 🤔 . I think I need to work a bit more on this.

raido commented:
This is now ready for feedback :)

bmish commented on tests/lib/rules/no-shadow-route-definition.js:
```diff
@@ -0,0 +1,594 @@
+      output: null,
+      errors: [
+        {
+          message: buildErrorMessage({
```
Could we check the string error message in at least one test? This `buildErrorMessage` is pretty complicated and we actually need to test that that generates the right string output too.

raido replied:
Test added.

raido commented on tests/lib/rules/no-shadow-route-definition.js:
```diff
@@ -0,0 +1,594 @@
+    "MyClass.route('blog', { path: '/blog' });",
+    "route.unrelatedFunction('blog', { path: 'blog' });",
+    "this.route.unrelatedFunction('blog', { path: 'blog' });",
+  ],
```
Test added.

raido commented on lib/rules/no-shadow-route-definition.js:
```diff
@@ -0,0 +1,158 @@
+  return routePath;
+}
+
+function getParentRoutesPaths(node, stack = []) {
```
Refactored.

raido commented:
I addressed feedback but this PR needs more work. Wildcard and params route detection logic is flawed.

Example:

```
this.route("first", { path: "/:first" });
this.route("second", { path: "/:second" });
```

Would not be detected as conflict but from router perspective it is. Just parameter name is different but URLs would be the same. Until I have fixed this, this cannot be merged.

raido commented:
I've implemented a solution for matching params and wildcards.
